File: tc-dotnet/Crc32.cs
using System.Buffers.Binary;

namespace TrueCryptReader;

/// <summary>
/// CRC-32 matching TrueCrypt's implementation (standard CRC-32/ISO-HDLC).
/// Polynomial: 0xEDB88320 (reflected 0x04C11DB7).
/// </summary>
public static class Crc32
{
    private static readonly uint[] Table = BuildTable();

    private static uint[] BuildTable()
    {
        var table = new uint[256];
        for (uint i = 0; i < 256; i++)
        {
            uint crc = i;
            for (int j = 0; j < 8; j++)
                crc = (crc & 1) != 0 ? (crc >> 1) ^ 0xEDB88320u : crc >> 1;
            table[i] = crc;
        }
        return table;
    }

    public static uint Compute(byte[] data, int offset, int length)
    {
        uint crc = 0xFFFFFFFF;
        for (int i = 0; i < length; i++)
            crc = Table[(crc ^ data[offset + i]) & 0xFF] ^ (crc >> 8);
        return crc ^ 0xFFFFFFFF;
    }

    public static uint Compute(ReadOnlySpan<byte> data)
    {
        uint crc = 0xFFFFFFFF;
        for (int i = 0; i < data.Length; i++)
            crc = Table[(crc ^ data[i]) & 0xFF] ^ (crc >> 8);
        return crc ^ 0xFFFFFFFF;
    }
}

File: tc-app/src-tauri/src/core/crc32.rs
/// CRC-32/ISO-HDLC matching TrueCrypt's implementation.
/// Polynomial: 0xEDB88320 (reflected 0x04C11DB7).

pub fn compute(data: &[u8]) -> u32 {
    let mut hasher = crc32fast::Hasher::new();
    hasher.update(data);
    hasher.finalize()
}

pub fn compute_slice(data: &[u8], offset: usize, length: usize) -> u32 {
    compute(&data[offset..offset + length])
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_known_values() {
        // CRC32 of empty input
        assert_eq!(compute(&[]), 0x0000_0000);
        // CRC32 of "123456789"
        assert_eq!(compute(b"123456789"), 0xCBF4_3926);
    }

    #[test]
    fn test_slice() {
        let data = b"xx123456789yy";
        assert_eq!(compute_slice(data, 2, 9), 0xCBF4_3926);
    }
}
